config/auth_local_metadata_service.go

```diff
@@ -0,0 +1,171 @@
+package config
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/databricks/databricks-sdk-go/logger"
+	"golang.org/x/oauth2"
+)
+
+// timeout to wait for local auth server to respond
+const metadataServiceTimeout = 10 * time.Second
+
+// Credentials provider that fetches a token from a locally running HTTP server
+//
+// The credentials provider will perform a GET request to the configured URL.
+//
+// The header "Metadata: true" will be added to the request, which must be
+// verified by server to prevent SSRF attacks
+//
+// The server is expected to return a JSON response with the following fields:
+//
+//	host: URL of the Databricks host to connect to.
+//	access_token: The requested access token.
+//	token_type: The type of token, which is a "Bearer" access token.
+//	expires_on: The timespan when the access token expires.
+type LocalMetadataServiceCredentials struct{}
+
+func (c LocalMetadataServiceCredentials) Name() string {
+	return "local-metadata-service"
+}
+
+func (c LocalMetadataServiceCredentials) Configure(ctx context.Context, cfg *Config) (func(*http.Request) error, error) {
+	if cfg.LocalMetadataServiceUrl == "" {
+		return nil, nil
+	}
+
+	parsedMetadataServiceUrl, err := url.Parse(cfg.LocalMetadataServiceUrl)
+	if err != nil {
+		return nil, fmt.Errorf("invalid auth server URL: %w", err)
+	}
+
+	// only allow localhost URLs
+	if parsedMetadataServiceUrl.Hostname() != "localhost" && parsedMetadataServiceUrl.Hostname() != "127.0.0.1" {
+		return nil, fmt.Errorf("invalid auth server URL: %s", cfg.LocalMetadataServiceUrl)
+	}
+
+	resp, err := makeRequest(parsedMetadataServiceUrl)
+	if err != nil {
+		return nil, nil
+	}
+
+	if resp == nil {
+		return nil, nil
+	}
+
+	if resp.Host == "" {
+		return nil, nil
+	}
+
+	if cfg.Host != "" && cfg.Host != resp.Host {
+		return nil, nil
+	}
+
+	cfg.Host = resp.Host
+
+	ts := metadataServiceTokenSource{
+		metadataServiceURL: parsedMetadataServiceUrl,
+		databricksHost:     cfg.Host,
+	}
+	return refreshableVisitor(&ts), nil
+}
+
+// makes a request to the server and returns the token
+func makeRequest(serverUrl *url.URL) (*serverResponse, error) {
```

Comment:
Make implies construction of the request, not performing the request.

Potential name: "performRequest".

Also, you can do:
```go
type metadataService struct {
  *url.URL
}

func (s metadataService) Get() (*metadata, error)
```

To avoid references to the literal URL and a global function.